src/python/pants/backend/go/module.py

```diff
@@ -202,5 +212,54 @@ async def run_go_resolve(targets: UnexpandedTargets, workspace: Workspace) -> Go
     return GoResolveGoal(exit_code=0)
 
 
+@dataclass(frozen=True)
+class DownloadExternalModuleRequest:
+    path: str
+    version: str
+
+
+@dataclass(frozen=True)
+class DownloadedExternalModule:
+    path: str
+    version: str
+    digest: Digest
+
+
+@rule
+async def download_external_module(
+    request: DownloadExternalModuleRequest,
+) -> DownloadedExternalModule:
+    result = await Get(
+        ProcessResult,
+        GoSdkProcess(
+            input_digest=EMPTY_DIGEST,
+            command=("mod", "download", "-json", f"{request.path}@{request.version}"),
+            description=f"Download external Go module at {request.path}@{request.version}.",
+            output_directories=("gopath",),
+        ),
+    )
+
+    # Decode the module metadata.
+    metadata = json.loads(result.stdout)
+
+    # Find the path within the digest where the source was downloaded. The path will have a sandbox-specific
+    # prefix that we need to strip down to the `gopath` path component.
+    absolute_source_path = metadata["Dir"]
+    gopath_index = absolute_source_path.index("gopath/")
+    source_path = absolute_source_path[gopath_index:]
+
+    source_digest = await Get(
+        Digest, DigestSubset(result.output_digest, PathGlobs([f"{source_path}/**"]))
```

Comment:
Should this error if the subsetting fails? `GlobMatchErrorBehavior.error` https://www.pantsbuild.org/docs/rules-api-file-system#pathglobs

Reply:
Yes, it should. Good point.